### ISSUE TITLE
Datagrid page auto-adjust when reducing result-size by filtering

### ISSUE THREAD

**Issue by TacoWannee** [Megabit/Blazorise#4281]
I am loading Datgrid data dynamically. If you go the last page of a big list, the page number is max for the data that is available.
When you enter a filter condition, the result size reduces, and so the number of available pages reduces too.
The page control is adapting to the new number of pages, but the data retrieved is from the backend is of an unreachable page and so results no records.

It would be nice, that there was a logic, that in this case the control detects after retrieving the backend result, that the page is out of reach, and does a new call for a fitting page.

The backend can't just return the last records that fit in because this will not match with the pages.

**Comment by David-Moreira** [comment]
I see. Point in case here is for `ReadData` feature, which is all about flexibility, we provide you with the `Datagrid` info, and you query your backend for the corresponding data.

So, I'm not sure whether we should or not support this. We'd need to be detecting if any **new** filter had been applied and intentionally be mutating the `Datagrid` Page request back to the initial page for you.
But If we think about it, It does make sense. 
Would we have any unitended behaviour with `MultipleSort` for example? I don't think so... What do you think @stsrki ?

For now an option would be for you to do this detection yourself, I believe the Datagrid will be flexible enough for this to work.
- Detect that a new filter has been applied and query page 1 of Data.
- Set the DataGrid's CurrentPage to 1. 

Have you tried this? Or can you?
We might try to apply this on our Datagrid demo page also.

**Comment by stsrki** [comment]
I don't think we need to do anything here. When ReadData is involved, the user should completely control how the data is retrieved, including the filtering. If the backend doesn't have any data when filtering, that is an expected result and behavior.

**Comment by David-Moreira** [comment]
> If the backend doesn't have any data when filtering, that is an expected result and behavior.

Just making sure, you understand the issue here, I'll re-explain by steps:

In a standard Datagrid with ReadData:
1 - User visits page 5 with no filters applied
2 - User now filters by Name = "David"
3 - There are 3 users named David in the backend, **HOWEVER** **ReadData** will send page 5 + Column Filter David which  in a standard readdata implementation will always result in empty results, this gives the false sense, that there's no David's in the backend.

What @TacoWannee has to do currently, is detect the filter change, and query David for page 1. And reset the paging.

The question here is if we, would take this as a feature meaning,
- ReadData would be smart enough to detect the filter and send a Page 1 + column filter David Request. 

Does this makes sense? I actually see benefits. @stsrki ?

**Comment by David-Moreira** [comment]
Thing is, you don't even get the paging.
Testing on our demo site. 
If you have no filter and type Name="d" while on page 3. You'll get results. 
While if you have no filter and on Page 4 and type Name="d" you'll get nothing at all. 

Then again, it might just be our **DEMO** custom ReadData implementation that should be better. I'll take a look later.
![image](https://user-images.githubusercontent.com/22283161/201985841-fd812b14-edfc-479d-9792-95994d820522.png)

**Comment by stsrki** [comment]
It seems to me that demo code is wrong

![image](https://user-images.githubusercontent.com/900302/201999536-d8c82865-98ea-4b41-a31f-c91c4f51bcae.png)

**Comment by David-Moreira** [comment]
Why do you say that bit is wrong? After data has been filtered it's skipping the elements from previous pages, and taking only the requested page, it seems to me?

**Comment by stsrki** [comment]
The reason is that when filtering starts, we assume that the result will be on the same page number, but most of the time it is not. After the filter is applied, the resulting data will be less than before, even empty data. So I think that we should automatically go to page 1 when any filter is done.

**Comment by David-Moreira** [comment]
Ahhh but that's what I was advocating for. If ReadData would be 'smart' to do it itself, go to page 1.

Or the developer should implement handling for this.
I'll try doing this second option on the demo and see if everything works correctly.

> For now an option would be for you to do this detection yourself, I believe the Datagrid will be flexible enough for this to work.

> Detect that a new filter has been applied and query page 1 of Data.
Set the DataGrid's CurrentPage to 1.

**Comment by stsrki** [comment]
Yeah, developers should handle it themselves. In the demo, we're just faking the backend API but basically, the same think will happen in the actual backend.

**Comment by stsrki** [comment]
I'm looking at different Datagrids from the competition. All of them go to page one when filtering.

**Comment by David-Moreira** [comment]
> I'm looking at different Datagrids from the competition. All of them go to page one when filtering.

With **ReadData**? But do they do that by being "smart"? Or by DEV ReadData implementation?
